lib/pbench/agent/tool_data_sink.py

```diff
@@ -1263,11 +1264,10 @@ def execute(self):
             self._tm_tracking = self.record_tms(tms)
             self._num_tms = len(self._tm_tracking.keys())
 
-            status = check_failed_tool_init(tms)
-            # Tell the entity that started us who we are, indicating if we
-            # are ready or not based on the status collected from TMs and
-            # proceed further accordingly.
-            started_msg = dict(kind="ds", action="startup", status=status)
+            # Tell the entity that started us that we are ready to accept
+            # requests, now that all Tool Meisters have all reported back as
+            # ready, too.
+            started_msg = dict(kind="ds", action="startup", status="success")
```

Comment:
This feels odd:  so the protocol is either that the TDS will declare "success" or it will send no message at all?  (There's no case where the TDS would detect an external failure and report it back as such?  Moreover, if the TDS itself fails, it doesn't try to report that?)  Is the TDS's own status encapsulated in its subscription to the client's channel (such that, if the client tries to send a message it will see that a failed TDS is no longer there?)?

Reply:
Yeah; right now we don't have the concept of "startup failed", so it always succeeds. However, if one or more tool init failed, that's recorded in a separate Redis key that'll be queried outside the TDS, removing the TDS entirely from that protocol.

The only way TDS wouldn't publish the message would be if we get taken out by an exception in `wait_for_initial_tms()` or `record_tms()`. We should probably try to publish a `"failure"` status message to the controller in that case instead of just logging the exception and closing the channel.